constantine/math_compiler/experimental/backends/cuda.nim

```diff
@@ -114,11 +130,133 @@ proc genFunctionType*(typ: GpuType, fn: string, fnArgs: string): string =
 proc genMemcpy(lhs, rhs, size: string): string =
   result = &"memcpy({lhs}, {rhs}, {size})"
 
+proc scanFunctions(ctx: var GpuContext, n: GpuAst) =
+  ## Iterates over the given function and checks for all `gpuCall` nodes. Any function
+  ## called in the scope is added to `fnTab`. This is a form of dead code elimination.
+  case n.kind
+  of gpuCall:
+    let fn = n.cName
+    if fn in ctx.allFnTab:
+      # Check if any of the parameters are pointers (otherwise non generic)
+      if fn notin ctx.fnTab: # function not known, add to `fnTab` (i.e. avoid code elimination)
+        let fnCalled = ctx.allFnTab[fn]
+        ctx.fnTab[fn] = fnCalled
+        # still "scan for functions", i.e. fill `fnTab` from inner calls
+        for ch in fnCalled:
+          ctx.scanFunctions(ch)
+      # else we don't do anything for this function
+    # Harvest functions from arguments to this call!
+    for arg in n.cArgs:
+      ctx.scanFunctions(arg)
+  else:
+    for ch in n:
+      ctx.scanFunctions(ch)
+
+proc getFieldType(t: GpuType, field: GpuAst): GpuType =
+  ## Returns the type of the field. `t` must be an object or generic instantiation.
+  ## `field` must be an ident.
+  doAssert field.kind == gpuIdent, "Field is not an ident: " & $field
+  doAssert t.kind in [gtObject, gtGenericInst]
+  let flds = if t.kind == gtObject: t.oFields
+               else: t.gFields
+  result = GpuType(kind: gtInvalid)
+  for f in flds:
+    if f.name == field.ident():
+      return f.typ
+
+proc getType(ctx: var GpuContext, arg: GpuAst, typeOfIndex = true): GpuType =
+  ## Tries to determine the underlying type of the AST.
+  ##
+  ## If `typeOfIndex` is `true`, we return the type of the index we access. Otherwise
+  ## we return the type of the array / pointer.
+  ##
+  ## NOTE: Do *not* rely on this for `mutable` or `implicit` fields of pointer types!
```

Comment:
We might want to auto-handle `HiddenDeref`

Reply:
What do you have in mind by auto-handle? We _are_ auto handling `HiddenDeref`. What we _might_ want to do in the future is to copy Nim's AST by explicitly differentiating between `gpuDeref` and `gpuHiddenDeref` (which doesn't exist yet).